packages/apps/reputation-oracle/server/src/modules/webhook/webhook.service.ts

```diff
@@ -40,47 +57,160 @@ export class WebhookService {
+    payload: object,
+    hash: string,
+    url: string,
+  ): Promise<void> {
+    let webhookEntity = new WebhookOutgoingEntity();
+    webhookEntity.payload = payload;
+    webhookEntity.hash = hash;
+    webhookEntity.url = url;
+    webhookEntity.status = WebhookOutgoingStatus.PENDING;
     webhookEntity.waitUntil = new Date();
     webhookEntity.retriesCount = 0;
 
-    webhookEntity = await this.webhookRepository.createUnique(webhookEntity);
+    webhookEntity =
+      await this.webhookOutgoingRepository.createUnique(webhookEntity);
 
     if (!webhookEntity) {
       throw new ControlledError(ErrorWebhook.NotCreated, HttpStatus.NOT_FOUND);
     }
   }
 
   /**
-   * Handles errors that occur during webhook processing.
-   * It logs the error and, based on retry count, updates the webhook status accordingly.
-   * @param webhookEntity - The entity representing the webhook data.
-   * @param error - The error object thrown during processing.
-   * @returns {Promise<void>} - Returns a promise that resolves when the operation is complete.
+   * Creates a tracking record for escrow completion in the repository.
+   * Sets initial status to 'PENDING'.
+   * @param {ChainId} chainId - The blockchain chain ID.
+   * @param {string} escrowAddress - The address of the escrow contract.
+   * @throws {ControlledError} If the tracking record cannot be created.
    */
-  public async handleWebhookError(
+  public async createEscrowCompletionTracking(
+    chainId: ChainId,
+    escrowAddress: string,
+  ): Promise<void> {
+    let escrowCompletionTrackingEntity = new EscrowCompletionTrackingEntity();
+    escrowCompletionTrackingEntity.chainId = chainId;
+    escrowCompletionTrackingEntity.escrowAddress = escrowAddress;
+    escrowCompletionTrackingEntity.status =
+      EscrowCompletionTrackingStatus.PENDING;
+    escrowCompletionTrackingEntity.waitUntil = new Date();
+    escrowCompletionTrackingEntity.retriesCount = 0;
+
+    escrowCompletionTrackingEntity =
+      await this.escrowCompletionTrackingRepository.createUnique(
+        escrowCompletionTrackingEntity,
+      );
+
+    if (!escrowCompletionTrackingEntity) {
+      throw new ControlledError(
+        ErrorEscrowCompletionTracking.NotCreated,
+        HttpStatus.NOT_FOUND,
+      );
+    }
+  }
+
+  /**
+   * Handles errors that occur while processing an incoming webhook.
+   * If retry count is below the maximum, increments retry count and reschedules; otherwise, marks as 'FAILED'.
+   * @param webhookEntity - The incoming webhook entity.
+   * @param failedReason - Reason for the failure.
+   */
+  public async handleWebhookIncomingError(
     webhookEntity: WebhookIncomingEntity,
+    failedReason: string,
   ): Promise<void> {
-    if (webhookEntity.retriesCount >= this.serverConfigService.maxRetryCount) {
-      webhookEntity.status = WebhookStatus.FAILED;
+    if (webhookEntity.retriesCount < this.serverConfigService.maxRetryCount) {
+      webhookEntity.waitUntil = new Date();
```

Comment:
@eugenvoronov it would be better to move that calculation to some function that can can also accept interval other than 1 second, so it's reusable and easy to have different values for different use-cases.

Also cron runs every 2 minutes, so using short backoff suggested above might not make sense: `(2 ** 5) * 1000` is 32 seconds, so maybe we should have some bigger step, e.g. 1 minute

Reply:
@dnechay I see that bachkoff was also implemented by @flopez7 here
https://github.com/humanprotocol/human-protocol/commit/651047c4e3bbb79beae257b7c067bd5a4eb3c9b7

Reply:
@eugenvoronov 
What was implemented by flopez was something different and based on the number of runs to skip, not time like here, even though it was using some kind of exponential calculation for number of "skips"

Here we have `waitUntil` which is time-based, so we should consider that.

Taking into account that it's not "jobs queue" (e.g. like SQS, where jobs rely on the thing called `VisibilityTimeout` and you can actually control when they are consumed) but cron that runs every 2nd minute, time-based backoff makes sense only if `waitUntil` is bigger than next run cycle, so as options we have:
1. keep just `retriesCount` and assume our backoff is `2 minutes` or less (depends on cron runs), which might be bad over time because cron takes time to run and if it runs for 2 minutes then eventually there will be no pauses (but this option might be enough for now)
2. keep `waitUntil` but use time higher value than "run window", i.e. either fixed backoff > 2 minutes or exponential that grows to > 2 minutes

Reply:
Yes, I've reviewed it, you are right  
I've implemented backoff for `waitUntil` here https://github.com/humanprotocol/human-protocol/blob/9e6c84ca57493463032624e4ed8e81ce7c6f960c/packages/apps/reputation-oracle/server/src/common/utils/cron.ts